File: sim/src/domain/FactoryPatterns.ts
/**
 * FactoryPatterns — les huit patterns d'usine du firmware d'origine
 * (miroir de flexseq/FactoryPatterns.h).
 *
 * L'original livre A1..A8 avec du contenu et B1..B8 vides
 * (`Gravity.ino:83-98`). FlexSeq les livrait tous vides, ce qui vidait de sens
 * la regle qui gele A1..A8 : geler huit emplacements vides ne donne rien.
 *
 * Le contenu tient sur 16 bits par pattern, bit i = step i, donc 16 octets de
 * PROGMEM cote AVR. Les steps 16 et au-dela n'existent pas chez l'original et
 * restent eteints.
 */
import type { PatternBank } from "./PatternBank.js";

export const FACTORY_PATTERN_COUNT = 8;
export const FACTORY_STEP_COUNT = 16;

const FACTORY: readonly number[] = [
  0x9111, 0x0810, 0x1249, 0xcccc, 0xeeee, 0x5454, 0x7fbf, 0xb733,
];

export function loadFactoryPatterns(bank: PatternBank): void {
  for (let index = 0; index < FACTORY_PATTERN_COUNT; ++index) {
    const pattern = bank.getPattern(index);
    if (!pattern) continue;
    pattern.clear();
    pattern.setLowStepMask(FACTORY[index] ?? 0);
  }
}

File: src/domain/FactoryPatterns.cpp
#include <flexseq/FactoryPatterns.h>

#ifdef __AVR__
#include <avr/pgmspace.h>
#endif

namespace flexseq {

namespace {

#ifdef __AVR__
const uint16_t kFactory[FACTORY_PATTERN_COUNT] PROGMEM = {
#else
const uint16_t kFactory[FACTORY_PATTERN_COUNT] = {
#endif
    0x9111,
    0x0810,
    0x1249,
    0xCCCC,
    0xEEEE,
    0x5454,
    0x7FBF,
    0xB733,
};

}  // namespace

uint16_t factoryStepMask(uint8_t index) {
    if (index >= FACTORY_PATTERN_COUNT) {
        return 0;
    }
#ifdef __AVR__
    return static_cast<uint16_t>(pgm_read_word_near(&kFactory[index]));
#else
    return kFactory[index];
#endif
}

void loadFactoryPatterns(PatternBank& bank) {
    for (uint8_t index = 0; index < FACTORY_PATTERN_COUNT; ++index) {
        Pattern* pattern = bank.getPattern(index);
        if (pattern == nullptr) {
            continue;
        }
        pattern->clear();
        pattern->setLowStepMask(factoryStepMask(index));
    }
}

}  // namespace flexseq
